packages/react-components/src/components/chart/hooks/useTrendCursorsEvents.ts

```diff
@@ -55,50 +55,58 @@ const useTrendCursorsEvents = ({
+      if ((ignoreHotKey || isInCursorAddModeRef.current) && graphicRef.current.length < MAX_TREND_CURSORS) {
+        if (isInSyncModeRef.current) {
+          const timestampInMs = calculateTimeStamp(posX, chartRef);
+          addTrendCursorsToSyncState({
+            groupId: groupId ?? '',
+            tcId: `trendCursor-${uuid()}`,
+            timestamp: timestampInMs,
+          });
+        } else {
+          const newTc = getNewTrendCursor({
+            size: sizeRef.current,
+            series: seriesRef.current,
+            x: posX,
+            chartRef,
+            visualization,
+          });
+
+          if (newTc) {
+            setGraphicRef.current([...graphicRef.current, newTc]);
+          }
+        }
+      }
+    },
+    [groupId, chartRef, sizeRef, seriesRef, visualization, addTrendCursorsToSyncState]
+  );
+
+  // shared delete function between the context menu and on click actions
+  const deleteTrendCursor = useCallback(
+    (toBeDeletedGraphicIndex: number) => {
       if (isInSyncModeRef.current) {
-        const timestampInMs = calculateTimeStamp(posX, chartRef);
-        addTrendCursorsToSyncState({
+        deleteTrendCursorsInSyncState({
           groupId: groupId ?? '',
-          tcId: `trendCursor-${uuid()}`,
-          timestamp: timestampInMs,
+          tcId: graphicRef.current[toBeDeletedGraphicIndex].id as string,
         });
       } else {
-        const newTc = getNewTrendCursor({
-          size: sizeRef.current,
-          series: seriesRef.current,
-          x: posX,
-          chartRef,
-          visualization,
-        });
-
-        if (newTc) {
-          setGraphicRef.current([...graphicRef.current, newTc]);
-        }
+        graphicRef.current[toBeDeletedGraphicIndex].$action = 'remove';
+        graphicRef.current[toBeDeletedGraphicIndex].children = []; // Echarts will throw error if children are not empty
+        chartRef.current?.setOption({ graphic: graphicRef.current });
+        graphicRef.current.splice(toBeDeletedGraphicIndex, 1);
+        setGraphicRef.current([...graphicRef.current]);
       }
-    }
-  };
-
-  // shared delete function between the context menu and on click actions
-  const deleteTrendCursor = (toBeDeletedGraphicIndex: number) => {
-    if (isInSyncModeRef.current) {
-      deleteTrendCursorsInSyncState({
-        groupId: groupId ?? '',
-        tcId: graphicRef.current[toBeDeletedGraphicIndex].id as string,
-      });
-    } else {
-      graphicRef.current[toBeDeletedGraphicIndex].$action = 'remove';
-      graphicRef.current[toBeDeletedGraphicIndex].children = []; // Echarts will throw error if children are not empty
-      chartRef.current?.setOption({ graphic: graphicRef.current });
-      graphicRef.current.splice(toBeDeletedGraphicIndex, 1);
-      setGraphicRef.current([...graphicRef.current]);
-    }
-  };
+    },
+    [groupId, graphicRef, chartRef, deleteTrendCursorsInSyncState]
```

Comment:
the point of having refs like graphicRef was to not have this re run when the graphic changes, the ref will hold the latest value and can be referred directly